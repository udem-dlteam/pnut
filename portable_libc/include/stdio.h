#ifndef _STDIO_H
#define _STDIO_H

#include "sys/types.h"
#include "stdarg.h"

#ifdef USE_STRUCT

typedef struct {
  int fd;
  char buf[1];
  char *string_output_buf;
  size_t string_output_buf_size;
  size_t string_output_len;
} FILE;

#else

typedef int FILE;

#endif

#ifdef PNUT_CC

FILE *stdin;
FILE *stdout;
FILE *stderr;

#else

extern FILE *stdin;
extern FILE *stdout;
extern FILE *stderr;

#endif

FILE *fopen(const char *pathname, const char *mode);
FILE *fdopen(int fd, const char *mode);
int fclose(FILE *stream);

int fputc(int c, FILE *stream);
size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);
int fputs(const char *s, FILE *stream);
int puts(const char *s);

size_t fread(void *buffer, size_t size, size_t count, FILE *stream);
int fseek( FILE* stream, long offset, int origin );
long ftell( FILE* stream );
int remove(const char *_Filename);

int vfprintf(FILE *stream, const char *format, va_list ap);
int fprintf(FILE *stream, const char *format VAR_ARGS);
int printf(const char *format VAR_ARGS);

int vsnprintf(char *str, size_t size, const char *format, va_list ap);
int snprintf(char *str, size_t size, const char *format VAR_ARGS);
int sprintf(char *str, const char *format VAR_ARGS);

#endif
