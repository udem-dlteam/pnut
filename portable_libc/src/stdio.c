#include "../include/stdio.h"

#include "../include/stdlib.h"
#include "../include/unistd.h"
#include "../include/fcntl.h"

FILE _standard_files[4];

FILE *stdin = _standard_files+0;
FILE *stdout = _standard_files+1;
FILE *stderr = _standard_files+2;

char *_string_output_buf;
size_t _string_output_buf_size;
size_t _string_output_len;

int _get_fd(FILE *f) {
  if (f == _standard_files+0) {
    return 0;
  } else if (f == _standard_files+1) {
    return 1;
  } else if (f == _standard_files+2) {
    return 2;
  } else {
    return *f;
  }
}

int fopen_flags(const char *mode) {
  int res = 0;
  while (*mode == 'b') ++mode; // Ignore binary mode

  if (*mode == 'r') {
    res = O_RDONLY;
  } else if (*mode == 'w') {
    res = O_WRONLY | O_CREAT | O_TRUNC;
  } else if (*mode == 'a') {
    res = O_WRONLY | O_CREAT | O_APPEND;
  } else {
    return 0;
  }

  mode += 1;
  while (*mode == 'b') ++mode; // Ignore binary mode (it can appear after r, w, a)

  if (*mode == '+') {
    res = res & ~(O_RDONLY | O_WRONLY) | O_RDWR;
  }

  return res;
}

FILE *fdopen(int fd, const char *mode) {
  FILE *result = malloc(sizeof(FILE));
  if (result) {
    *result = fd;
  }
  return result;
}

FILE *fopen(const char *pathname, const char *mode) {
  int fd = open(pathname, fopen_flags(mode), 0666); // 0666 is the default mode
  if (fd == -1) {
    return 0;
  } else {
    return fdopen(fd, mode);
  }
}

int fclose(FILE *stream) {
  int fd = _get_fd(stream);
  free(stream);
  if (close(fd) == 0) {
    return 0;
  } else {
    return EOF;
  }
}

int fseek(FILE* stream, long offset, int origin) {
  return lseek(_get_fd(stream), offset, origin);
}

off_t ftell(FILE * stream) {
  return lseek (_get_fd(stream), 0, SEEK_CUR);
}

size_t fread(void *data, size_t size, size_t count, FILE *stream) {
  return read(_get_fd(stream), data, size * count) / size;
}

int remove(const char *filename) {
  return unlink(filename);
}

char _output_buf[1];

int fputc(int c, FILE *stream) {
  int fd = _get_fd(stream);
  if (fd == -1) {
    if (_string_output_len+1 < _string_output_buf_size) {
      _string_output_buf[_string_output_len] = c;
      ++_string_output_len;
    }
  } else {
    _output_buf[0] = c;
    write(fd, _output_buf, 1);
  }
  return c;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  const char *p = ptr;
  size_t n = size * nmemb;
  while (n) {
    fputc(*p, stream);
    ++p;
    --n;
  }
  return nmemb;
}

int _fputstr(const char *str, FILE *stream) {
  int result = 0;
  while (*str) {
    fputc(*str, stream);
    ++str;
    ++result;
  }
  return result;
}

int fputs(const char *s, FILE *stream) {
  _fputstr(s, stream);
  // fputc('\n', stream);
  return 0;
}

int puts(const char *s) {
  return fputs(s, stdout);
}

int fflush(FILE *stream) {
  return 0; // no buffering so nothing to do
}

#define SIZEOF_NUM_BUF 100
char _num_buf[SIZEOF_NUM_BUF];

char *_int_to_str(int n, int base, int width, int force_0, int force_plus) {

  char *out = _num_buf + SIZEOF_NUM_BUF;
  int neg = n < 0;
  int has_sign = neg || force_plus;
  char *digits = "0123456789abcdef";

  if (!neg) n = -n;

  *--out = 0;

  while (n <= -base) {
    *--out = digits[-(n % base)];
    n /= base;
  }

  *--out = digits[-(n % base)];

  if (width >= SIZEOF_NUM_BUF) width = SIZEOF_NUM_BUF-1;

  if (force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width - has_sign) {
      *--out = force_0 ? '0' : ' ';
    }
  }

  if (has_sign) *--out = neg ? '-' : '+';

  if (!force_0) {
    while (_num_buf + SIZEOF_NUM_BUF - out <= width) {
      *--out = ' ';
    }
  }

  return out;
}

int vfprintf(FILE *stream, const char *format, va_list ap) {

  char c;
  int result = 0;
  int base;
  int width;
  int force_0;

  while ((c = *format++)) {
    if (c == '%' && (c = *format++) != '%') {
      width = 0;
      force_0 = 0;
      if (c == '0') {
        force_0 = 1;
        c = *format++;
      }
      while (c >= '0' && c <= '9') {
        width = width*10 + (c - '0');
        c = *format++;
      }
      while (c == 'l' || c == 'L') {
        c = *format++;
      }
      if (c == 'c') {
        fputc(va_arg(ap, int), stream);
        result += 1;
      } else if (c == 'd' || c == 'u' || c == 'o' || c == 'x') {
        base = c == 'x' ? 16 : c == 'o' ? 8 : 10;
        result += _fputstr(_int_to_str(va_arg(ap, int), base, width, force_0, 0), stream);
      } else if (c == 's') {
        result += _fputstr(va_arg(ap, char*), stream);
      } else {
        fputc('?', stream);
        fputc('?', stream);
        fputc('?', stream);
        result += 3;
      }
    } else {
      fputc(c, stream);
      result += 1;
    }
  }

  return result;
}

int fprintf(FILE *stream, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vfprintf(stream, format, ap);
  va_end(ap);

  return result;
}

int printf(const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vfprintf(stdout, format, ap);
  va_end(ap);

  return result;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {

  int result;

  _standard_files[3] = -1; /* mark as string FILE */
  _string_output_buf = str;
  _string_output_buf_size = size;
  _string_output_len = 0;
  result = vfprintf(_standard_files+3, format, ap);
  _string_output_buf[_string_output_len] = 0; /* null terminate string */

  return result;
}

int snprintf(char *str, size_t size, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vsnprintf(str, size, format, ap);
  va_end(ap);

  return result;
}

int sprintf(char *str, const char *format VAR_ARGS) {

  va_list ap;
  int result;

  va_start(ap, format);
  result = vsnprintf(str, 999999999, format, ap);
  va_end(ap);

  return result;
}
