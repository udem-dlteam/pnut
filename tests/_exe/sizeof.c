#include <stdio.h>

#ifdef PNUT_CC
// When bootstrapping pnut, intptr_t is not defined.
// On 64 bit platforms, intptr_t is a long long int.
// On 32 bit (including shells) platforms, intptr_t is an int.
#if defined(PNUT_EXE_64)
#define PTR_MUL 8
#else
#define PTR_MUL 4
#endif
#else
#define PTR_MUL 4
#endif


void putint(int n) {
  if (n < 0) {
    putchar('-');
    n = -n;
  }
  if (n > 9) {
    putint(n / 10);
  }
  putchar('0' + n % 10);
}

struct S0 {
  char a;
}; // 1 byte

struct S1 {
  int a;      // 4 bytes
  int b;      // 4 bytes
}; // 8 bytes

struct S2 {
  int a;
  char b;
  // 3 bytes padding
  int c;
}; // 12 bytes

struct S3 {
  char a;
  char b[2];
  // 1 bytes padding
  int c;
}; // 8 bytes

struct S4 {
  char a;       // 1 byte
  int b[0];     // 4 * 0 = 0 bytes
}; // 1 byte, largest member is int (1)

struct S5 {
  char b[4];
}; // 4 bytes, largest member is char (1)

struct S6 {
  int a;
}; // 4 bytes, largest member is int (4)

union U0 {
  char a;       // 1 byte
  int b;        // 4 bytes
}; // 4 bytes

union U1 {
  char a;       // 1 byte
  int b;        // 4 bytes
  char c[5];    // 5 bytes
}; // 8 bytes because aligned with width of int (4)

union U2 {
  char a[3];    // 3 bytes
  int b;        // 4 bytes
}; // 4 bytes because aligned with width of int (4)

union U3 {
  char a[5];    // 5 bytes
  char b;       // 1 byte
}; // 5 bytes because aligned with width of char (1)

union U4 {
  char a[5];    // 5 bytes
  int b;        // 4 bytes
}; // 8 bytes, because aligned with width of int (4)

union U5 {
  char a[5];    // 5 bytes
  struct S5 b;  // 4 bytes
}; // 5 bytes because aligned with width of char (1)

union U6 {
  char a[5];
  struct S6 b;
}; // 5 bytes

union U7 {
  char a[9];
  struct S5 b[1]; // 4 bytes
}; // 9 bytes because aligned with width of char (1)

union U8 {
  char a[2][3];   // 6 bytes
  struct S5 b[1]; // 4 bytes
}; // 6 bytes because aligned with width of char (1)

union U9 {
  char a[2][3];   // 6 bytes
  struct S4 b[1]; // 1 byte, but largest member is int (array of size 0)
}; // 6 bytes because aligned with width of int (4)

union U10 {
  char a[2][3];   // 6 bytes
  struct S5 b[1]; // 4 bytes
  int  c[1][1][1][1]; // 4 bytes
}; // 8 bytes because aligned with width of int (4)

struct ArHdr {
  char ar_name[16];
  char ar_date[12];
  char ar_uid[6];
  char ar_gid[6];
  char ar_mode[8];
  char ar_size[10];
  char ar_fmag[2];
};

int main() {
  int a[10] = { 0, 1, 2, 3 };
  int b[]   = { 0, 1, 2, 3 }; // Infer size from initializer (4 elements)
  putint(sizeof(char));       putchar('\n'); // No division needed because sizeof(char) is always 1
  putint(sizeof(int));        putchar('\n');
  putint(sizeof(int[10]));    putchar('\n');
  putint(sizeof(int[10][2])); putchar('\n');
  putint(sizeof(a));          putchar('\n'); // sizeof (expr)
  putint(sizeof a);           putchar('\n'); // sizeof expr
  putint(sizeof(b));          putchar('\n'); // sizeof (expr)
  putint(sizeof b);           putchar('\n'); // sizeof expr
  putint(sizeof((void *) a) / PTR_MUL); putchar('\n'); // sizeof (cast_expr)

  // Testing struct/union size and alignment
  putint(sizeof(struct S0)); putchar('\n');
  putint(sizeof(struct S1)); putchar('\n');
  putint(sizeof(struct S2)); putchar('\n');
  putint(sizeof(struct S3)); putchar('\n');
  putint(sizeof(struct S4)); putchar('\n');

  putint(sizeof(union U0)); putchar('\n');
  putint(sizeof(union U1)); putchar('\n');
  putint(sizeof(union U3)); putchar('\n');
  putint(sizeof(union U3[1])); putchar('\n');
  putint(sizeof(union U3[2])); putchar('\n');
  putint(sizeof(union U3[5])); putchar('\n');
  putint(sizeof(union U4)); putchar('\n');
  putint(sizeof(union U5)); putchar('\n');
  putint(sizeof(union U6)); putchar('\n');
  putint(sizeof(union U7)); putchar('\n');
  putint(sizeof(union U8)); putchar('\n');
  putint(sizeof(union U9)); putchar('\n');
  putint(sizeof(union U10)); putchar('\n');
  putint(sizeof(struct ArHdr)); putchar('\n');

  return 0;
}
