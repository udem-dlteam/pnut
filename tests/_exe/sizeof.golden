1
4
40
80
40
40
16
16
1
1
8
12
8
4
4
8
5
5
10
25
8
5
8
9
6
8
8
60
